File: RootTools2/src/main/java/com/rarnu/tools/neo/xposed/ads/FuckVideo.java
package com.rarnu.tools.neo.xposed.ads;

import android.content.Context;
import android.content.Loader;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import com.rarnu.tools.neo.xposed.XpUtils;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.callbacks.XC_LoadPackage;
import org.json.JSONObject;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by rarnu on 11/18/16.
 */
public class FuckVideo {

    public static void fuckVideo(XC_LoadPackage.LoadPackageParam loadPackageParam) {
        XpUtils.findAndHookMethod("com.miui.videoplayer.ads.DynamicAd", loadPackageParam.classLoader, "replaceList", List.class, String.class, new XC_MethodHook() {
            protected void beforeHookedMethod(MethodHookParam paramAnonymousMethodHookParam)
                    throws Throwable {
                paramAnonymousMethodHookParam.args[0] = null;
                paramAnonymousMethodHookParam.args[1] = null;
            }
        });
        XpUtils.findAndHookMethod("com.video.ui.view.AdView", loadPackageParam.classLoader, "getAdsBlock", Context.class, XC_MethodReplacement.returnConstant(null));
        Class<?> clsCallback = XpUtils.findClass(loadPackageParam.classLoader, "com.video.ui.idata.SyncServiceHelper$Callback");
        if (clsCallback != null) {
            XpUtils.findAndHookMethod("com.video.ui.idata.SyncServiceHelper", loadPackageParam.classLoader, "fetchAds", Context.class, clsCallback, XC_MethodReplacement.returnConstant(null));
        }
        XpUtils.findAndHookMethod("com.video.ui.idata.iDataORM", loadPackageParam.classLoader, "getBooleanValue", Context.class, String.class, boolean.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                String key = (String) param.args[1];
                if (key.equals("debug_mode") || key.equals("show_first_ads") || key.equals("ads_show_homekey") || key.equals("startup_ads_loop") || key.equals("app_upgrade_splash")) {
                    param.setResult(false);
                }
            }

            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                String key = (String) param.args[1];
                if (key.equals("debug_mode") || key.equals("show_first_ads") || key.equals("ads_show_homekey") || key.equals("startup_ads_loop") || key.equals("app_upgrade_splash")) {
                    param.setResult(false);
                }
            }
        });
        XpUtils.findAndHookMethod("com.video.ui.idata.iDataORM", loadPackageParam.classLoader, "getStringValue", Context.class, String.class, String.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                String key = (String) param.args[1];
                if (key.equals("startup_ads")) {
                    param.setResult(null);
                }
            }

            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                String key = (String) param.args[1];
                if (key.equals("startup_ads")) {
                    param.setResult(null);
                }
            }
        });
        XpUtils.findAndHookMethod("com.video.ui.idata.iDataORM", loadPackageParam.classLoader, "getBooleanValue", Context.class, String.class, boolean.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                String key = (String) param.args[1];
                if (key.equals("show_title_ads") || key.equals("show_channel_title_ads")) {
                    param.setResult(false);
                }
            }

            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                String key = (String) param.args[1];
                if (key.equals("show_title_ads") || key.equals("show_channel_title_ads")) {
                    param.setResult(false);
                }
            }
        });
        XpUtils.findAndHookMethod("com.video.ui.idata.iDataORM", loadPackageParam.classLoader, "enabledAds", Context.class, XC_MethodReplacement.returnConstant(false));
        Class<?> clsAdListener = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.systemAdSolution.splashAd.IAdListener");
        if (clsAdListener != null) {
            XpUtils.findAndHookMethod("com.miui.ad.sdk.api.RemoteSystemSplashAdService", loadPackageParam.classLoader, "requestSystemSplashAd", clsAdListener, XC_MethodReplacement.returnConstant(false));
            XpUtils.findAndHookMethod("com.miui.ad.sdk.api.SystemSplashAd", loadPackageParam.classLoader, "requestAd", Context.class, clsAdListener, XC_MethodReplacement.returnConstant(null));
            XpUtils.findAndHookMethod("com.miui.ad.sdk.api.SystemSplashAd", loadPackageParam.classLoader, "requestAd", clsAdListener, XC_MethodReplacement.returnConstant(null));
        }

        XpUtils.findAndHookMethod("com.miui.videoplayer.model.OnlineUri", loadPackageParam.classLoader, "supportFrontAD", XC_MethodReplacement.returnConstant(false));
        XpUtils.findAndHookMethod("com.miui.videoplayer.model.OnlineUri", loadPackageParam.classLoader, "supportPauseAD", XC_MethodReplacement.returnConstant(false));
        XpUtils.findAndHookMethod("com.miui.videoplayer.model.OnlineUri", loadPackageParam.classLoader, "supportCornerAD", XC_MethodReplacement.returnConstant(false));
        XpUtils.findAndHookMethod("com.miui.videoplayer.model.OnlineUri", loadPackageParam.classLoader, "skipAllAD", XC_MethodReplacement.returnConstant(true));
        XpUtils.findAndHookMethod("com.miui.videoplayer.model.OnlineUri", loadPackageParam.classLoader, "skipSDKAD", XC_MethodReplacement.returnConstant(true));
        XpUtils.findAndHookMethod("com.miui.videoplayer.model.OnlineUri", loadPackageParam.classLoader, "getMiAdFlag", XC_MethodReplacement.returnConstant(-1));

        XpUtils.findAndHookMethod("com.miui.videoplayer.ads.AdsContainer", loadPackageParam.classLoader, "setCornerAd", XC_MethodReplacement.returnConstant(null));
        XpUtils.findAndHookMethod("com.miui.videoplayer.ads.AdsContainer", loadPackageParam.classLoader, "enableOfflineAds", XC_MethodReplacement.returnConstant(false));

        XpUtils.findAndHookMethod("com.miui.videoplayer.videoview.VideoViewContainer", loadPackageParam.classLoader, "playAd", XC_MethodReplacement.returnConstant(null));
        Class<?> clsVideoView = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.videoplayer.videoview.IVideoView");
        if (clsVideoView != null) {
            XpUtils.findAndHookMethod("com.miui.videoplayer.videoview.VideoViewContainer", loadPackageParam.classLoader, "playRealVideo", clsVideoView, boolean.class, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    param.args[1] = true;   // skipAd
                }
            });
        }
        XpUtils.findAndHookMethod("com.miui.videoplayer.videoview.VideoViewContainer", loadPackageParam.classLoader, "prepareRealVideoView", boolean.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                param.args[0] = true;  // haveAd
            }
        });

        XpUtils.findAndHookMethod("com.miui.videoplayer.ads.AdsService", loadPackageParam.classLoader, "doLaunch", String.class, XC_MethodReplacement.returnConstant(null));
        XpUtils.findAndHookMethod("com.miui.videoplayer.videoview.MiAdsVideoView", loadPackageParam.classLoader, "haveAd", XC_MethodReplacement.returnConstant(false));
        XpUtils.findAndHookMethod("com.xiaomi.miui.ad.listeners.impl.AdEventListenerImpl", loadPackageParam.classLoader, "onAdRequest", String.class, JSONObject.class, XC_MethodReplacement.returnConstant(null));
        XpUtils.findAndHookMethod("com.xiaomi.miui.ad.listeners.impl.AdEventListenerImpl", loadPackageParam.classLoader, "onAdClicked", String.class, JSONObject.class, XC_MethodReplacement.returnConstant(null));

        Class<?> clsBlock = XpUtils.findClass(loadPackageParam.classLoader, "com.tv.ui.metro.model.Block");
        Class<?> clsPool = XpUtils.findClass(loadPackageParam.classLoader, "android.support.v7.widget.RecyclerView$RecycledViewPool");
        if (clsBlock != null && clsPool != null) {
            XpUtils.findAndHookConstructor("com.video.ui.view.BlockAdapter", loadPackageParam.classLoader, Context.class, clsBlock, new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    removeAds(param);
                }
            });
            XpUtils.findAndHookConstructor("com.video.ui.view.BlockAdapter", loadPackageParam.classLoader, Context.class, clsBlock, clsPool, new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    removeAds(param);
                }
            });
            XpUtils.findAndHookMethod("com.video.ui.view.BlockAdapter", loadPackageParam.classLoader, "addGroup", clsBlock, new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    removeAds(param);
                }
            });
            XpUtils.findAndHookMethod("com.video.ui.view.block.PortBlockView", loadPackageParam.classLoader, "initUI", clsBlock, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    Object block = param.args[0];
                    Class<?> clsBlock = block.getClass();
                    Field fBlocks = clsBlock.getDeclaredField("blocks");
                    fBlocks.setAccessible(true);
                    ArrayList blocks = (ArrayList) fBlocks.get(block);
                    for (int i = blocks.size() - 1; i >= 0; i--) {
                        Object o = blocks.get(i);
                        Field fUI = o.getClass().getSuperclass().getDeclaredField("ui_type");
                        Object oUI = fUI.get(o);
                        Method mId = oUI.getClass().getMethod("id");
                        int id = (int) mId.invoke(oUI);
                        if (id == 101 || id == 10001 || id == 282 || id == 257 || id == 221 ||
                                id == 501 || id == 502 || id == 503 || id == 601 | id == 602 || id == 603 || id == 604) {
                            blocks.remove(i);
                        }
                    }
                }
            });
        }

        if (clsBlock != null) {
            XpUtils.findAndHookMethod("com.video.ui.view.ListFragment", loadPackageParam.classLoader, "setBlockView", clsBlock, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    Object o = param.args[0];
                    Field fBlocks = o.getClass().getDeclaredField("blocks");
                    fBlocks.setAccessible(true);
                    fBlocks.set(o, null);
                    Field fFooters = o.getClass().getDeclaredField("footers");
                    fFooters.setAccessible(true);
                    fFooters.set(o, null);
                }
            });
        }

        Class<?> clsVideo = XpUtils.findClass(loadPackageParam.classLoader, "com.tv.ui.metro.model.VideoItem");
        if (clsVideo != null) {
            XpUtils.findAndHookMethod("com.video.ui.view.DetailFragment", loadPackageParam.classLoader, "updateVideo", clsVideo, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    Object oVideo = param.args[0];
                    Field fBlocks = oVideo.getClass().getDeclaredField("blocks");
                    fBlocks.setAccessible(true);
                    fBlocks.set(oVideo, null);
                    Field fHeaders = oVideo.getClass().getDeclaredField("headers");
                    fHeaders.setAccessible(true);
                    fHeaders.set(oVideo, null);
                }
            });
        }

        XpUtils.findAndHookMethod("com.video.ui.view.DetailFragment", loadPackageParam.classLoader, "onCreateView", LayoutInflater.class, ViewGroup.class, Bundle.class, new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                Object fragment = param.thisObject;
                Field fR1 = fragment.getClass().getDeclaredField("relative_region");
                Field fR2 = fragment.getClass().getDeclaredField("headers_region");
                FrameLayout r1 = (FrameLayout) fR1.get(fragment);
                FrameLayout r2 = (FrameLayout) fR2.get(fragment);
                r1.setVisibility(View.GONE);
                r2.setVisibility(View.GONE);
            }
        });
        XpUtils.findAndHookMethod("com.video.ui.view.DetailFragment", loadPackageParam.classLoader, "checkAdsPresentVisibility", View.class, XC_MethodReplacement.returnConstant(null));
        XpUtils.findAndHookMethod("com.video.ui.view.ListFragment", loadPackageParam.classLoader, "checkAdsPresentVisibility", View.class, XC_MethodReplacement.returnConstant(null));

        Class<?> clsGenericBlock = XpUtils.findClass(loadPackageParam.classLoader, "com.tv.ui.metro.model.GenericBlock");
        if (clsGenericBlock != null) {
            XpUtils.findAndHookMethod("com.video.ui.view.user.MyVideoFragment", loadPackageParam.classLoader, "onLoadFinished", Loader.class, clsGenericBlock, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    param.args[1] = null;
                }
            });
        }
    }

    private static void removeAds(XC_MethodHook.MethodHookParam param) {
        try {
            Class<?> clsThis = param.thisObject.getClass();
            Field fBlockRootArrayList = clsThis.getDeclaredField("mBlockRootArrayList");
            fBlockRootArrayList.setAccessible(true);
            ArrayList mBlockRootArrayList = (ArrayList) fBlockRootArrayList.get(param.thisObject);
            for (int i = mBlockRootArrayList.size() - 1; i >= 0; i--) {
                Object b = mBlockRootArrayList.get(i);
                Field fUI = b.getClass().getSuperclass().getDeclaredField("ui_type");
                fUI.setAccessible(true);
                Object oUI = fUI.get(b);
                Method mId = oUI.getClass().getMethod("id");
                int id = (int) mId.invoke(oUI);
                if (id == 101 || id == 10001 || id == 282 || id == 257 || id == 221 ||
                        id == 501 || id == 502 || id == 503 || id == 601 | id == 602 || id == 603 || id == 604) {
                    mBlockRootArrayList.remove(i);
                }
            }
        } catch (Exception e) {
        }
    }
}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/xposed/ads/FuckVideo.kt
@file:Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS", "DEPRECATION", "Duplicates")

package com.rarnu.tools.neo.xposed.ads

import android.content.Context
import android.content.Loader
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.rarnu.xfunc.*
import de.robv.android.xposed.XC_MethodHook
import org.json.JSONObject
import java.util.*

/**
 * Created by rarnu on 11/18/16.
 */
object FuckVideo {

    fun fuckVideo(pkg: XposedPkg) {
        pkg.findClass("com.miui.videoplayer.ads.DynamicAd").findMethod("replaceList", List::class.java, String::class.java).hook {
            before {
                args[0] = null
                args[1] = null
            }
        }
        pkg.findClass("com.video.ui.view.AdView").findMethod("getAdsBlock", Context::class.java).hook { replace { result = null } }

        val clsCallback = pkg.findClass("com.video.ui.idata.SyncServiceHelper\$Callback")
        if (clsCallback != null) {
            pkg.findClass("com.video.ui.idata.SyncServiceHelper").findMethod("fetchAds", Context::class.java, clsCallback).hook { replace { result = null } }
        }
        pkg.findClass("com.video.ui.idata.iDataORM").findMethod("getBooleanValue", Context::class.java, String::class.java, java.lang.Boolean.TYPE).hook {
            before {
                val key = args[1] as String
                if (key == "debug_mode" || key == "show_first_ads" || key == "ads_show_homekey" || key == "startup_ads_loop" || key == "app_upgrade_splash") {
                    result = false
                }
            }
            after {
                val key = args[1] as String
                if (key == "debug_mode" || key == "show_first_ads" || key == "ads_show_homekey" || key == "startup_ads_loop" || key == "app_upgrade_splash") {
                    result = false
                }
            }
        }
        pkg.findClass("com.video.ui.idata.iDataORM").findMethod("getStringValue", Context::class.java, String::class.java, String::class.java).hook {
            before {
                val key = args[1] as String
                if (key == "startup_ads") {
                    result = null
                }
            }
            after {
                val key = args[1] as String
                if (key == "startup_ads") {
                    result = null
                }
            }
        }
        pkg.findClass("com.video.ui.idata.iDataORM").findMethod("getBooleanValue", Context::class.java, String::class.java, java.lang.Boolean.TYPE).hook {
            before {
                val key = args[1] as String
                if (key == "show_title_ads" || key == "show_channel_title_ads") {
                    result = false
                }
            }
            after {
                val key = args[1] as String
                if (key == "show_title_ads" || key == "show_channel_title_ads") {
                    result = false
                }
            }
        }
        pkg.findClass("com.video.ui.idata.iDataORM").findMethod("enabledAds", Context::class.java).hook { replace { result = false } }

        val clsAdListener = pkg.findClass("com.miui.systemAdSolution.splashAd.IAdListener")
        if (clsAdListener != null) {
            pkg.findClass("com.miui.ad.sdk.api.RemoteSystemSplashAdService").findMethod("requestSystemSplashAd", clsAdListener).hook { replace { result = false } }
            pkg.findClass("com.miui.ad.sdk.api.SystemSplashAd").apply {
                findMethod("requestAd", Context::class.java, clsAdListener).hook { replace { result = null } }
                findMethod("requestAd", clsAdListener).hook { replace { result = null } }
            }
        }

        pkg.findClass("com.miui.videoplayer.model.OnlineUri").apply {
            findMethod("supportFrontAD").hook { replace { result = false } }
            findMethod("supportPauseAD").hook { replace { result = false } }
            findMethod("supportCornerAD").hook { replace { result = false } }
            findMethod("skipAllAD").hook { replace { result = true } }
            findMethod("skipSDKAD").hook { replace { result = true } }
            findMethod("getMiAdFlag").hook { replace { result = -1 } }
        }

        pkg.findClass("com.miui.videoplayer.ads.AdsContainer").apply {
            findMethod("setCornerAd").hook { replace { result = null } }
            findMethod("enableOfflineAds").hook { replace { result = false } }
        }
        pkg.findClass("com.miui.videoplayer.videoview.VideoViewContainer").findMethod("playAd").hook { replace { result = null } }

        val clsVideoView = pkg.findClass("com.miui.videoplayer.videoview.IVideoView")
        if (clsVideoView != null) {
            pkg.findClass("com.miui.videoplayer.videoview.VideoViewContainer").findMethod("playRealVideo", clsVideoView, java.lang.Boolean.TYPE).hook {
                before {
                    args[1] = true   // skipAd
                }
            }
        }

        pkg.findClass("com.miui.videoplayer.videoview.VideoViewContainer").findMethod("prepareRealVideoView", java.lang.Boolean.TYPE).hook {
            before {
                args[0] = true  // haveAd
            }
        }

        // TODO:
        pkg.findClass("com.miui.videoplayer.ads.AdsService").findMethod("doLaunch", String::class.java).hook { replace { result = null } }
        pkg.findClass("com.miui.videoplayer.videoview.MiAdsVideoView").findMethod("haveAd").hook { replace { result = false } }
        pkg.findClass("com.xiaomi.miui.ad.listeners.impl.AdEventListenerImpl").apply {
            findMethod("onAdRequest", String::class.java, JSONObject::class.java).hook { replace { result = null } }
            findMethod("onAdClicked", String::class.java, JSONObject::class.java).hook { replace { result = null } }
        }

        val clsBlock = pkg.findClass("com.tv.ui.metro.model.Block")
        val clsPool = pkg.findClass("android.support.v7.widget.RecyclerView\$RecycledViewPool")
        if (clsBlock != null && clsPool != null) {
            pkg.findClass("com.video.ui.view.BlockAdapter").apply {
                findConstructor(Context::class.java, clsBlock).hook {
                    after { removeAds(this) }
                }
                findConstructor(Context::class.java, clsBlock, clsPool).hook {
                    after { removeAds(this) }
                }
                findMethod("addGroup", clsBlock).hook {
                    after { removeAds(this) }
                }
            }

            pkg.findClass("com.video.ui.view.block.PortBlockView").findMethod("initUI", clsBlock).hook {
                before {
                    val block = args[0]
                    val clsBlock2 = block.javaClass
                    val fBlocks = clsBlock2.getDeclaredField("blocks").apply { isAccessible = true }
                    val blocks = fBlocks.get(block) as ArrayList<*>
                    blocks.indices.reversed().forEach {
                        val o = blocks[it]
                        val fUI = o.javaClass.superclass.getDeclaredField("ui_type")
                        val oUI = fUI.get(o)
                        val mId = oUI.javaClass.getMethod("id")
                        val id = mId.invoke(oUI) as Int
                        if (id == 101 || id == 10001 || id == 282 || id == 257 || id == 221 ||
                                id == 501 || id == 502 || id == 503 || (id == 601) or (id == 602) || id == 603 || id == 604) {
                            blocks.removeAt(it)
                        }
                    }
                }
            }
        }

        if (clsBlock != null) {
            pkg.findClass("com.video.ui.view.ListFragment").findMethod("setBlockView", clsBlock).hook {
                before {
                    val o = args[0]
                    val fBlocks = o.javaClass.getDeclaredField("blocks").apply { isAccessible = true }
                    fBlocks.set(o, null)
                    val fFooters = o.javaClass.getDeclaredField("footers").apply { isAccessible = true }
                    fFooters.set(o, null)
                }
            }
        }

        val clsVideo = pkg.findClass("com.tv.ui.metro.model.VideoItem")
        if (clsVideo != null) {
            pkg.findClass("com.video.ui.view.DetailFragment").findMethod("updateVideo", clsVideo).hook {
                before {
                    val oVideo = args[0]
                    val fBlocks = oVideo.javaClass.getDeclaredField("blocks").apply { isAccessible = true }
                    fBlocks.set(oVideo, null)
                    val fHeaders = oVideo.javaClass.getDeclaredField("headers").apply { isAccessible = true }
                    fHeaders.set(oVideo, null)
                }
            }
        }

        pkg.findClass("com.video.ui.view.DetailFragment").apply {
            findMethod("onCreateView", LayoutInflater::class.java, ViewGroup::class.java, Bundle::class.java).hook {
                after {
                    val fragment = thisObject
                    val fR1 = fragment.javaClass.getDeclaredField("relative_region")
                    val fR2 = fragment.javaClass.getDeclaredField("headers_region")
                    val r1 = fR1.get(fragment) as FrameLayout
                    val r2 = fR2.get(fragment) as FrameLayout
                    r1.visibility = View.GONE
                    r2.visibility = View.GONE
                }
            }
            findMethod("checkAdsPresentVisibility", View::class.java).hook { replace { result = null } }
        }

        pkg.findClass("com.video.ui.view.ListFragment").findMethod("checkAdsPresentVisibility", View::class.java).hook { replace { result = null } }

        val clsGenericBlock = pkg.findClass("com.tv.ui.metro.model.GenericBlock")
        if (clsGenericBlock != null) {
            pkg.findClass("com.video.ui.view.user.MyVideoFragment").findMethod("onLoadFinished", Loader::class.java, clsGenericBlock).hook {
                before {
                    args[1] = null
                }
            }
        }

        pkg.findClass("com.miui.systemAdSolution.landingPage.LandingPageService").findMethod("init", Context::class.java).hook { replace { result = null } }

    }

    private fun removeAds(param: XC_MethodHook.MethodHookParam) {
        try {
            val clsThis = param.thisObject?.javaClass
            val fBlockRootArrayList = clsThis?.getDeclaredField("mBlockRootArrayList")?.apply { isAccessible = true }
            val mBlockRootArrayList = fBlockRootArrayList?.get(param.thisObject) as ArrayList<*>

            mBlockRootArrayList.indices.reversed().forEach {
                val b = mBlockRootArrayList[it]
                val fUI = b.javaClass.superclass.getDeclaredField("ui_type").apply { isAccessible = true }
                val oUI = fUI.get(b)
                val mId = oUI.javaClass.getMethod("id")
                val id = mId.invoke(oUI) as Int
                if (id == 101 || id == 10001 || id == 282 || id == 257 || id == 221 ||
                        id == 501 || id == 502 || id == 503 || (id == 601) or (id == 602) || id == 603 || id == 604) {
                    mBlockRootArrayList.removeAt(it)
                }
            }
        } catch (e: Exception) {
        }

    }
}
